src/reducers/profile-view.js

```diff
@@ -205,6 +205,64 @@ function viewOptionsPerThread(state: ThreadViewOptions[] = [], action: Action) {
         ...state.slice(threadIndex + 1),
       ];
     }
+    case 'CHANGE_IMPLEMENTATION_FILTER': {
+      const {
+        transformedThread,
+        threadIndex,
+        previousImplementation,
+        implementation,
+      } = action;
+
+      if (previousImplementation === implementation) {
+        return state;
+      }
+
+      // This CallNodePath may need to be updated twice.
+      let selectedCallNodePath = state[threadIndex].selectedCallNodePath;
+
+      if (
+        implementation !== 'combined' &&
+        previousImplementation !== 'combined'
+      ) {
+        // Restore the CallNodePath back to an unfiltered state before re-filtering
+        // it on the next implementation.
+        selectedCallNodePath = Transforms.restoreAllFunctionsInCallNodePath(
+          transformedThread,
+          previousImplementation,
+          selectedCallNodePath
+        );
+      }
+
+      if (implementation === 'combined') {
+        // Restore the full CallNodePaths
+        selectedCallNodePath = Transforms.restoreAllFunctionsInCallNodePath(
+          transformedThread,
+          previousImplementation,
+          selectedCallNodePath
+        );
```

Comment:
This call is exactly the same as the one in the previous if branch. I think the two if branches should be combined into one `if (previousImplementation !== 'combined') {`. The else branch can then become `if (implementation !== 'combined') {`. I think that would not only be shorter but also clearer, and logically equivalent.